### PR TITLE
Populate DATABASEOID cache before CREATE INDEX CONCURRENTLY

### LINKED ISSUE #4675: CREATE INDEX CONCURRENTLY might lead to self-deadlock
```sql
CREATE TABLE brintest_3 (a text, b text, c text, d text);
SELECT create_reference_table('brintest_3');
CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);
```

-----

Some useful info for debugging:

```sql
┌──────────────────┬─────────────┬──────────────┬───────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ blocked_locktype │ blocked_pid │ blocking_pid │   lock_mode   │                                          blocked_statement                                          │                                current_statement_in_blocking_process                                │
├──────────────────┼─────────────┼──────────────┼───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ virtualxid       │      22619  │       22653  │ ExclusiveLock │ CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);                                      │ CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102008 ON public.brintest_3_102008 USING btree (a )  │
│ virtualxid       │      22653  │       22619  │ ShareLock     │ CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102008 ON public.brintest_3_102008 USING btree (a )  │ CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);                                      │
└──────────────────┴─────────────┴──────────────┴───────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

```c
// backtrace of 22619

(gdb) bt
#0  0x00007f3a96d4d9f3 in __epoll_wait_nocancel () at ../sysdeps/unix/syscall-template.S:84
#1  0x00000000009390bf in WaitEventSetWaitBlock (set=0x21cef30, cur_timeout=1000, occurred_events=0x21cf390, nevents=3) at latch.c:1295
#2  0x0000000000938f7c in WaitEventSetWait (set=0x21cef30, timeout=1000, occurred_events=0x21cf390, nevents=3, wait_event_info=100663296) at latch.c:1247
#3  0x00007f3a94496cef in RunDistributedExecution (execution=0x21ce5b0) at executor/adaptive_executor.c:2269
#4  0x00007f3a944951d3 in ExecuteTaskListExtended (executionParams=0x21ce520) at executor/adaptive_executor.c:995
#5  0x00007f3a94494f83 in ExecuteUtilityTaskList (utilityTaskList=0x2371c28, localExecutionSupported=false) at executor/adaptive_executor.c:880
#6  0x00007f3a9446e1a2 in ExecuteDistributedDDLJob (ddlJob=0x2263eb8) at commands/utility_hook.c:946
#7  0x00007f3a9446dc98 in ProcessUtilityInternal (pstmt=0x22637f0, queryString=0x2127750 "CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x21286f8, completionTag=0x7fff30d0c9c0) at commands/utility_hook.c:673
#8  0x00007f3a9446cfc6 in multi_ProcessUtility (pstmt=0x2128418, queryString=0x2127750 "CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x21286f8, completionTag=0x7fff30d0c9c0) at commands/utility_hook.c:251
#9  0x00007f3a9454394a in ColumnarProcessUtility (pstmt=0x2128418, queryString=0x2127750 "CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x21286f8, completionTag=0x7fff30d0c9c0) at ../columnar/cstore_tableam.c:1348
#10 0x00007f3a8ec64541 in pgss_ProcessUtility (pstmt=0x2128418, queryString=0x2127750 "CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x21286f8, qc=0x7fff30d0c9c0) at pg_stat_statements.c:1156
#11 0x0000000000972505 in ProcessUtility (pstmt=0x2128418, queryString=0x2127750 "CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x21286f8, qc=0x7fff30d0c9c0) at utility.c:520
#12 0x0000000000971330 in PortalRunUtility (portal=0x22078b0, pstmt=0x2128418, isTopLevel=true, setHoldSnapshot=false, dest=0x21286f8, qc=0x7fff30d0c9c0)
    at pquery.c:1157
#13 0x0000000000971554 in PortalRunMulti (portal=0x22078b0, isTopLevel=true, setHoldSnapshot=false, dest=0x21286f8, altdest=0x21286f8, qc=0x7fff30d0c9c0)
    at pquery.c:1303
#14 0x00000000009709dd in PortalRun (portal=0x22078b0, count=9223372036854775807, isTopLevel=true, run_once=true, dest=0x21286f8, altdest=0x21286f8,
    qc=0x7fff30d0c9c0) at pquery.c:779
#15 0x000000000096a60c in exec_simple_query (query_string=0x2127750 "CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);") at postgres.c:1239
#16 0x000000000096e96d in PostgresMain (argc=1, argv=0x21be6f8, dbname=0x21be500 "regression", username=0x21be4d8 "postgres") at postgres.c:4315
#17 0x00000000008b5ca3 in BackendRun (port=0x21b96a0) at postmaster.c:4536
#18 0x00000000008b541c in BackendStartup (port=0x21b96a0) at postmaster.c:4220
#19 0x00000000008b1851 in ServerLoop () at postmaster.c:1739
#20 0x00000000008b1080 in PostmasterMain (argc=5, argv=0x2122130) at postmaster.c:1412
#21 0x00000000007b0914 in main (argc=5, argv=0x2122130) at main.c:210
```

```c
// backtrace of 22653

(gdb) bt
#0  0x00007f3a96d4d9f3 in __epoll_wait_nocancel () at ../sysdeps/unix/syscall-template.S:84
#1  0x00000000009390bf in WaitEventSetWaitBlock (set=0x21d0788, cur_timeout=-1, occurred_events=0x7fff30d0b1c0, nevents=1) at latch.c:1295
#2  0x0000000000938f7c in WaitEventSetWait (set=0x21d0788, timeout=-1, occurred_events=0x7fff30d0b1c0, nevents=1, wait_event_info=50331654) at latch.c:1247
#3  0x00000000009385fd in WaitLatchOrSocket (latch=0x7f3a8e454d64, wakeEvents=33, sock=-1, timeout=-1, wait_event_info=50331654) at latch.c:428
#4  0x0000000000938461 in WaitLatch (latch=0x7f3a8e454d64, wakeEvents=33, timeout=0, wait_event_info=50331654) at latch.c:368
#5  0x000000000096015b in ProcSleep (locallock=0x21976a8, lockMethodTable=0xcc8c00 <default_lockmethod>) at proc.c:1292
#6  0x000000000094e8e3 in WaitOnLock (locallock=0x21976a8, owner=0x21c00c8) at lock.c:1859
#7  0x000000000094d474 in LockAcquireExtended (locktag=0x7fff30d0b5d0, lockmode=5, sessionLock=false, dontWait=false, reportMemoryError=true,
    locallockp=0x0) at lock.c:1101
#8  0x000000000094c9de in LockAcquire (locktag=0x7fff30d0b5d0, lockmode=5, sessionLock=false, dontWait=false) at lock.c:752
#9  0x0000000000953acd in VirtualXactLock (vxid=..., wait=true) at lock.c:4546
#10 0x00000000006adf11 in WaitForOlderSnapshots (limitXmin=4581, progress=true) at indexcmds.c:445
#11 0x00000000006b010c in DefineIndex (relationId=34064, stmt=0x22a7248, indexRelationId=34949, parentIndexId=0, parentConstraintId=0,
    is_alter_table=false, check_rights=true, check_not_in_use=true, skip_build=false, quiet=false) at indexcmds.c:1547
#12 0x0000000000974599 in ProcessUtilitySlow (pstate=0x2288aa0, pstmt=0x2288988,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, qc=0x7fff30d0c9c0) at utility.c:1519
#13 0x00000000009735e5 in standard_ProcessUtility (pstmt=0x2288988,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, qc=0x7fff30d0c9c0) at utility.c:1069
#14 0x00007f3a9446d897 in ProcessUtilityInternal (pstmt=0x2288988,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, completionTag=0x7fff30d0c9c0) at commands/utility_hook.c:558
#15 0x00007f3a9446cfc6 in multi_ProcessUtility (pstmt=0x2128528,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, completionTag=0x7fff30d0c9c0) at commands/utility_hook.c:251
#16 0x00007f3a9454394a in ColumnarProcessUtility (pstmt=0x2128528,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, completionTag=0x7fff30d0c9c0) at ../columnar/cstore_tableam.c:1348
#17 0x00007f3a8ec64541 in pgss_ProcessUtility (pstmt=0x2128528,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, qc=0x7fff30d0c9c0) at pg_stat_statements.c:1156
#18 0x0000000000972505 in ProcessUtility (pstmt=0x2128528,
    queryString=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ",
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x2128808, qc=0x7fff30d0c9c0) at utility.c:520
#19 0x0000000000971330 in PortalRunUtility (portal=0x2207ea0, pstmt=0x2128528, isTopLevel=true, setHoldSnapshot=false, dest=0x2128808, qc=0x7fff30d0c9c0)
    at pquery.c:1157
#20 0x0000000000971554 in PortalRunMulti (portal=0x2207ea0, isTopLevel=true, setHoldSnapshot=false, dest=0x2128808, altdest=0x2128808, qc=0x7fff30d0c9c0)
    at pquery.c:1303
#21 0x00000000009709dd in PortalRun (portal=0x2207ea0, count=9223372036854775807, isTopLevel=true, run_once=true, dest=0x2128808, altdest=0x2128808,
    qc=0x7fff30d0c9c0) at pquery.c:779
#22 0x000000000096a60c in exec_simple_query (
    query_string=0x2127750 "CREATE  INDEX CONCURRENTLY  brin_test_temp_idx_102505 ON public.brintest_3_102505 USING btree (a ) ") at postgres.c:1239
#23 0x000000000096e96d in PostgresMain (argc=1, argv=0x21be640, dbname=0x21be500 "regression", username=0x21be4d8 "postgres") at postgres.c:4315
#24 0x00000000008b5ca3 in BackendRun (port=0x21b96a0) at postmaster.c:4536
#25 0x00000000008b541c in BackendStartup (port=0x21b96a0) at postmaster.c:4220
#26 0x00000000008b1851 in ServerLoop () at postmaster.c:1739
#27 0x00000000008b1080 in PostmasterMain (argc=5, argv=0x2122130) at postmaster.c:1412
#28 0x00000000007b0914 in main (argc=5, argv=0x2122130) at main.c:210
```

### PR DESCRIPTION
**Pull request by pykello**
Fixes #4675 

In https://github.com/citusdata/citus/issues/4090 we avoided local execution for `CREATE INDEX CONCURRENTLY` to avoid a crash. Part of that PR was starting a new transaction in the backend processing the utility commands before sending the `CREATE INDEX CONCURRENTLY` over the shard connection. This was to reset the xmin of the utility command backend to 0 so the last phase of CIC doesn't hang waiting for the utility command backend when calling `WaitForOlderSnapshots()`.

But if the `DATABASEOID` cache is not populated, then when we open new connections in the adaptive executor, the catalog table look-up in `get_database_name()` can change the xmin. This can happen for example right after creating the Citus extension.

To avoid this, we call `get_database_name()` before starting the transaction to populate the `DATABASEOID` cache and avoid the xmin change.

### CODE REVIEW

**Review by onurctirtir — COMMENTED** [review]
I know this issue only reproduces right after `create extension` but maybe we can still have regression tests from https://github.com/citusdata/citus/pull/4687#discussion_r586188623 in this pr too

**Comment by onderkalaci** [comment]
> I know this issue only reproduces right after create extension but maybe we can still have regression tests from #4687 (comment) in this pr too

I'd even be in favor of DROP EXTENSION/CREATE EXTENSION on the single_node.sql file, and follow the example in the issue, which can repro the problem always. And, I verified that your commit fixes it
```SQL
CREATE TABLE brintest_3 (a text, b text, c text, d text);
SELECT create_reference_table('brintest_3');
CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);
```

**Review by onderkalaci — APPROVED** [review]
I think this looks good, and fixes the issue we have seen. I think we should find a way to add the test

 And, this also explains why we only saw this right after `CREATE EXTENSION`. For example, if I do the following, it works fine on master as well when the cache is not empty:

```SQL

CREATE EXTENSION citus;
CREATE TABLE brintest_3 (a text, b text, c text, d text);
SELECT create_reference_table('brintest_3');
SELECT * FROM citus_tables;
-- works fine after the above SELECT
CREATE INDEX CONCURRENTLY brin_test_temp_idx ON brintest_3(a);
```

File: src/backend/distributed/commands/utility_hook.c [hunk 1]
```diff
@@ -910,6 +910,19 @@ ExecuteDistributedDDLJob(DDLJob *ddlJob)
...
+			 * the xmin of current backend to change. Then the last phase
+			 * of CREATE INDEX CONCURRENTLY, which is in a separate backend,
+			 * will hang waiting for our backend and result in a deadlock.
+			 *
+			 * We populate the cache before starting the next transaction to
+			 * avoid this. Most of the metadata has already been resolved in
+			 * planning phase, we only need to lookup metadata needed for
+			 * connection establishment.
+			 */
+			(void) get_database_name(MyDatabaseId);
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:924]: could also be `CurrentDatabaseName` to populate the Citus cache

**Comment by pykello** [comment]
@onderkalaci @onurctirtir I tried adding suggested tests, and they weren't straight-forward and will take some time. Given my limited time in Citus can one of you work on adding them in a separate PR?

**Pull request merged**